# 1. Importing Libraries and Data

## 1.1. Libraries

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

## 1.2. Data

In [2]:
# Create Path
path = r'C:\Users\M de Villiers\Desktop\Data Analytics\Acheivement 4\InstaCart_Basket_Analysis_10_01_2022'

In [3]:
# Import Data: Orders
df_ords = pd.read_csv(os.path.join(path, '02_Data', '02_Prepared_Data', 'orders_wrangled1.csv'), index_col = False)

In [4]:
# Import Data: Products
df_prods = pd.read_csv(os.path.join(path, '02_Data', '01_Original_Data', 'products.csv'), index_col = False)

# 2. Exercise Walk-Through

## 2.1. Mixed Data Types

In [5]:
# Create Test Data Frame
df_test = pd.DataFrame()

In [6]:
df_test['mix'] = ['a', 'b', 1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
# Check for mixed data types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [9]:
df_test['mix'] = df_test['mix'].astype('str')

## 2.2. Missing Values

In [10]:
# Check for missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
# create df of missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [12]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [13]:
df_prods.shape

(49693, 5)

In [14]:
# create new df without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [15]:
df_prods_clean

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

In [16]:
df_prods_clean.shape

(49677, 5)

## 2.3. Duplicates

In [17]:
# Finding duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [18]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [19]:
df_prods_clean.shape

(49677, 5)

In [20]:
# address duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [21]:
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

In [22]:
df_prods_clean_no_dups.shape

(49672, 5)

In [23]:
#export data frame: checked products
df_prods_clean_no_dups.to_csv(os.path.join(path, '02_Data','02_Prepared_Data', 'products_checked.csv'), index = False)

# 3. Task 4.5

### Question 2

In [25]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

#### The maximum amount under prices seems unrealistic. I doubt anything in a grocery store will cost 99999 dollars (assuming it is an American company). It must be a mistake. Perhaps 99.99 or 999.99 dollars (if they sell very expensive kitchen appliances such as food processors or coffee machines), but definitely not 99999 dollars.

#### Product_id also has a higher max than count but this could be due to duplicates that were removed. The duplicates perhaps had a unique id even though they were duplicates.

### Question 3

In [26]:
# Check for mixed data types in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

#### It appears that there are no columns with mixed data types in the orders data frame.

### Questions 4

#### The code executed in Question 3 did not print any problematic columns (columns with mixed data types), thus no columns to be fixed in this regard.

### Question 5

In [27]:
# Check for missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#### There are 206209 missing or 0 values in the days_since_prior_order column. The value is probably 0 when a customer or user orders for the first time (thus there will be no prior order). It could also be that some people order only a few hours apart. Since a full day = 24 hours, it could be possible that some people ordered items late in the afternoon the one day and perhaps the next day they ordered something earlier in the day (e.g. morning) and then the system would show 0 days since prior order seeing as 24 hours has not gone by yet.

In [ ]:
# Crosstab
xtab = pd.crosstab(final_df1['age_range'], final_df['spending_flag'], dropna = False)

### Question 6

In [27]:
# Check missing values for a pattern
df_ords_check = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [28]:
df_ords_check.head(30)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0     2539329        1             1                  2                  8   
11    2168274        2             1                  2                 11   
26    1374495        3             1                  1                 14   
39    3343014        4             1                  6                 11   
45    2717275        5             1                  3                 12   
50    2086598        6             1                  5                 18   
54    2565571        7             1                  3                  9   
75     600894        8             1                  6                  0   
79     280530        9             1                  1                 17   
83    1224907       10             1                  2                 14   
89    2411567       11             1                  0                 11   
97    1483446       12             1                  1                 12   
103   2618231       13             1                  6                 12   
116   2906263       14             1                  2                 18   
130   2514324       15             1                  5                 13   
153   1273676       16             1                  2                  8   
160   1737705       17             1                  2                 13   
201   2780889       18             1                  6                 18   
208   1321974       19             1                  6                 13   
218    947722       20             1                  1                 10   
223   2625565       21             1                  1                 10   
257   2300537       22             1                  3                  7   
273   1792521       23             1                  5                 14   
278    611214       24             1                  3                 20   
297   2610576       25             1                  1                 14   
301   3229714       26             1                  2                 15   
314    361493       27             1                  3                  9   
396   1138968       28             1                  0                 14   
421   2064090       29             1                  5                 11   
440    375303       30             1                  1                 13   

     days_since_prior_order  
0                       NaN  
11                      NaN  
26                      NaN  
39                      NaN  
45                      NaN  
50                      NaN  
54                      NaN  
75                      NaN  
79                      NaN  
83                      NaN  
89                      NaN  
97                      NaN  
103                     NaN  
116                     NaN  
130                     NaN  
153                     NaN  
160                     NaN  
201                     NaN  
208                     NaN  
218                     NaN  
223                     NaN  
257                     NaN  
273                     NaN  
278                     NaN  
297                     NaN  
301                     NaN  
314                     NaN  
396                     NaN  
421                     NaN  
440                     NaN

#### As can be seen from the 30 entries printed above, these were all first orders (order number = 1) and thus there would be no prior order (and 0 days since prior order). These are not "missing" values or problematic. The only thing to do here is to impute the missing values (NaN) with "0" since there has been no prior order.

In [29]:
df_ords

order_id  user_id  order_number  order_day_of_week  \
0         2539329        1             1                  2   
1         2398795        1             2                  3   
2          473747        1             3                  3   
3         2254736        1             4                  4   
4          431534        1             5                  4   
...           ...      ...           ...                ...   
3421078   2266710   206209            10                  5   
3421079   1854736   206209            11                  4   
3421080    626363   206209            12                  1   
3421081   2977660   206209            13                  1   
3421082    272231   206209            14                  6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

In [30]:
#define new data frame and imputing NaN with 0
df_ords_clean = df_ords
df_ords_clean.fillna(0, inplace =True)

In [31]:
df_ords_clean

order_id  user_id  order_number  order_day_of_week  \
0         2539329        1             1                  2   
1         2398795        1             2                  3   
2          473747        1             3                  3   
3         2254736        1             4                  4   
4          431534        1             5                  4   
...           ...      ...           ...                ...   
3421078   2266710   206209            10                  5   
3421079   1854736   206209            11                  4   
3421080    626363   206209            12                  1   
3421081   2977660   206209            13                  1   
3421082    272231   206209            14                  6   

         order_hour_of_day  days_since_prior_order  
0                        8                     0.0  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

### Question 7

In [32]:
# Finding duplicates
df_ords_dups = df_ords[df_ords.duplicated()]

In [33]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [34]:
df_ords_dups.shape

(0, 6)

#### There seems to be no duplicates.

### Question 8

#### If there were any duplicates in the dataframe I would have used the following method to address the duplicates: df_ords_no_dups = df_ords.drop_duplicates() since this would have deleted or dropped all duplicate entries / records.

### Question 9

In [35]:
df_ords_clean.to_csv(os.path.join(path, '02_Data','02_Prepared_Data', 'orders_cleaned.csv'), index = False)

#### I exported the products dataframe at the end of the exercise walkthrough earlier.